drivers/mfd/mfd_max31790.c

```diff
@@ -0,0 +1,153 @@
+
+LOG_MODULE_REGISTER(max_max31790, CONFIG_MFD_LOG_LEVEL);
+
+struct max31790_config {
+	struct i2c_dt_spec i2c;
+};
+
+struct max31790_data {
+};
+
+static void max31790_set_globalconfiguration_i2cwatchdog(uint8_t *destination, uint8_t value)
+{
+	uint8_t length = MAX37190_GLOBALCONFIGURATION_I2CWATCHDOG_LENGTH;
+	uint8_t pos = MAX37190_GLOBALCONFIGURATION_I2CWATCHDOG_POS;
+
+	*destination &= ~GENMASK(pos + length - 1, pos);
+	*destination |= FIELD_PREP(GENMASK(pos + length - 1, pos), value);
+}
+
+int max31790_read_register_uint8(const struct device *dev, uint8_t address, uint8_t *value)
+{
+	const struct max31790_config *config = dev->config;
+	int result;
+
+	result = i2c_reg_read_byte_dt(&config->i2c, address, value);
+	if (result != 0) {
+		LOG_ERR("unable to read register 0x%02X, error %i", address, result);
+	}
+
+	LOG_DBG("read value 0x%02X from register 0x%02X", *value, address);
+
+	return result;
+}
+
+int max31790_read_register_uint16(const struct device *dev, uint8_t address, uint16_t *value)
+{
+	const struct max31790_config *config = dev->config;
+	int result;
+
+	result = i2c_write_read_dt(&config->i2c, &address, sizeof(address), value, sizeof(*value));
+	if (result != 0) {
+		LOG_ERR("unable to read from address 0x%02X, error %i", address, result);
+	}
+
+	*value = sys_be16_to_cpu(*value);
+
+	LOG_DBG("read value 0x%04X from address 0x%02X", *value, address);
+
+	return result;
+}
+
+int max31790_write_register_uint8(const struct device *dev, uint8_t address, uint8_t value)
+{
+	const struct max31790_config *config = dev->config;
+	int result;
+
+	LOG_DBG("writing value 0x%02X to register 0x%02X", value, address);
+	result = i2c_reg_write_byte_dt(&config->i2c, address, value);
+	if (result != 0) {
+		LOG_ERR("unable to write register 0x%02X, error %i", address, result);
+	}
+
+	return result;
+}
+
+int max31790_write_register_uint16(const struct device *dev, uint8_t address, uint16_t value)
+{
+	const struct max31790_config *config = dev->config;
+	int result;
+	uint8_t buffer[3];
+
+	buffer[0] = address;
+	sys_put_be16(value, buffer + 1);
+
+	LOG_DBG("writing value 0x%04X to address 0x%02X", value, address);
+	result = i2c_write_dt(&config->i2c, buffer, sizeof(buffer));
+	if (result != 0) {
+		LOG_ERR("unable to write to address 0x%02X, error %i", address, result);
+	}
+
+	return result;
+}
+
+static int max31790_init(const struct device *dev)
+{
+	const struct max31790_config *config = dev->config;
+	int result;
+	uint8_t reg_value;
+
+	if (!i2c_is_ready_dt(&config->i2c)) {
+		LOG_ERR("I2C device not ready");
+		return -ENODEV;
+	}
+
+	reg_value = 0;
+	reg_value &= ~MAX37190_GLOBALCONFIGURATION_STANDBY_BIT;
+	reg_value |= MAX37190_GLOBALCONFIGURATION_RESET_BIT;
+	reg_value |= MAX37190_GLOBALCONFIGURATION_BUSTIMEOUT_BIT;
+	reg_value &= ~MAX37190_GLOBALCONFIGURATION_OSCILLATORSELECTION_BIT;
+	max31790_set_globalconfiguration_i2cwatchdog(&reg_value, 0);
+	reg_value &= ~MAX37190_GLOBALCONFIGURATION_I2CWATCHDOGSTATUS_BIT;
```

Comment:
Some of this seems a little confusing (ANDing masks that are known not to be set).
Could this be simplified to a single value?

Reply:
I actually prefer setting or resetting the bits explicitly, because it explains what should be achieved. Otherwise there would be just a magic number, where one would have to dig into the datasheet to figure out what it means.
And performance wise I think it won't make a difference, I'm pretty sure the compiler will optimize the hell out of this sequence.

Reply:
Short, but then you loose the information that you do not want to set standby, oscillator selection and the watchdog. This information is then implicit and hidden.